<a href="https://colab.research.google.com/github/noely-b/House-Prices/blob/main/House_Prices_Checkpoint_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Zillow's "Home Values" dataset will be used in the project. The business problem that I'm attempting to resolve is to determine housing price and whether it'll be profitable to invest in them. Cleaning the data means that the data would be more accurate.

In [ ]:
#Import pandas and upload file
import pandas as pd
df = pd.read_csv('House Pricing Over Time.csv')

In [ ]:
#Verify the data's heading
df.head()

RegionID  SizeRank       RegionName StateName    1/31/2015    2/28/2015  \
0    394913         1     New York, NY        NY  405297.6670  405494.4639   
1    753899         2  Los Angeles, CA        CA  499076.8857  500816.1083   
2    394463         3      Chicago, IL        IL  192187.3867  192496.1100   
3    394514         4       Dallas, TX        TX  172153.8541  173654.2457   
4    394692         5      Houston, TX        TX  175277.2678  176958.6412   

     3/31/2015    4/30/2015    5/31/2015    6/30/2015  ...    4/30/2024  \
0  405959.9343  406562.6324  407768.0443  408937.5187  ...  647851.0011   
1  501186.8886  501331.7462  501954.2631  503193.3853  ...  911031.4988   
2  193029.0076  193551.2233  194417.6632  195118.8740  ...  311972.0067   
3  175138.1641  176502.9147  177931.3386  179505.2541  ...  372240.1988   
4  178615.0447  180097.5127  181506.6356  182837.9781  ...  306473.9927   

     5/31/2024    6/30/2024    7/31/2024    8/31/2024    9/30/2024  \
0  652477.0774  655536.1337  658263.3867  661979.7505  665977.9427   
1  916863.7169  920777.1781  925109.6435  930243.5852  936603.5177   
2  314053.1343  315253.9468  315976.4932  316881.6315  317866.6565   
3  372383.0861  371723.0017  370790.4746  370143.4467  369946.9588   
4  307007.4231  306889.4104  306526.8178  306309.4155  306247.0647   

    10/31/2024   11/30/2024   12/31/2024    1/31/2025  
0  669477.2350  672228.2408  673999.8383  675063.9760  
1  941773.6741  946228.3038  950210.0621  950740.4946  
2  318826.5634  319846.1184  321070.7257  322389.5161  
3  369841.8006  369555.4606  369035.0884  368515.9005  
4  306049.2835  305716.9513  305539.4578  305480.0785  

[5 rows x 125 columns]

In [ ]:
#See info from the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 894 entries, 0 to 893
Columns: 125 entries, RegionID to 1/31/2025
dtypes: float64(121), int64(2), object(2)
memory usage: 873.2+ KB


In [ ]:
#Describe the data
df.describe()

RegionID    SizeRank      1/31/2015      2/28/2015      3/31/2015  \
count     894.000000  894.000000     854.000000     854.000000     855.000000   
mean   412446.539150  462.268456  150783.202021  151252.944616  151954.179890   
std     77730.850434  268.416053   82982.029360   83688.992489   84433.959047   
min    394297.000000    1.000000   51367.814890   51089.157170   51188.625440   
25%    394547.250000  231.250000  101222.654900  101325.869175  101531.997450   
50%    394795.500000  460.500000  127869.200150  128153.532350  128473.171500   
75%    395044.750000  689.750000  172313.372200  172441.658100  172941.883150   
max    753929.000000  939.000000  774847.626900  784297.142300  789912.831400   

           4/30/2015      5/31/2015      6/30/2015      7/31/2015  \
count     856.000000     856.000000     857.000000     857.000000   
mean   152685.651141  153553.848853  154349.882036  155173.912416   
std     84933.896110   85452.094806   85920.470114   86453.900678   
min     51087.573560   50592.713710   50021.330590   49944.236190   
25%    101943.601750  102460.371850  103069.837000  103441.181100   
50%    129012.713250  129495.130700  129881.895300  130183.774800   
75%    173604.251300  174840.822175  175429.530600  176246.157100   
max    792572.428700  795334.411400  800136.453800  809309.379000   

           8/31/2015  ...     4/30/2024     5/31/2024     6/30/2024  \
count     857.000000  ...  8.940000e+02  8.940000e+02  8.940000e+02   
mean   155998.761914  ...  2.745421e+05  2.761255e+05  2.769601e+05   
std     87038.795459  ...  1.733260e+05  1.742478e+05  1.745009e+05   
min     50006.847530  ...  5.025394e+04  5.131431e+04  5.180810e+04   
25%    103996.208600  ...  1.691034e+05  1.703707e+05  1.709575e+05   
50%    130801.724700  ...  2.249000e+05  2.272253e+05  2.283484e+05   
75%    178219.441000  ...  3.258355e+05  3.273618e+05  3.280330e+05   
max    821439.217200  ...  1.500225e+06  1.517969e+06  1.532970e+06   

          7/31/2024     8/31/2024     9/30/2024    10/31/2024    11/30/2024  \
count  8.940000e+02  8.940000e+02  8.940000e+02  8.940000e+02  8.940000e+02   
mean   2.772719e+05  2.775254e+05  2.778279e+05  2.783259e+05  2.788188e+05   
std    1.746016e+05  1.748047e+05  1.754167e+05  1.758304e+05  1.759435e+05   
min    5.164894e+04  5.118189e+04  5.082892e+04  5.060702e+04  4.987629e+04   
25%    1.712436e+05  1.718653e+05  1.717518e+05  1.717072e+05  1.718808e+05   
50%    2.286165e+05  2.290341e+05  2.289249e+05  2.295252e+05  2.305109e+05   
75%    3.296642e+05  3.299217e+05  3.304091e+05  3.312606e+05  3.315559e+05   
max    1.547438e+06  1.560217e+06  1.572113e+06  1.582338e+06  1.594660e+06   

         12/31/2024     1/31/2025  
count  8.940000e+02  8.940000e+02  
mean   2.796469e+05  2.804037e+05  
std    1.758092e+05  1.755394e+05  
min    4.927014e+04  4.898139e+04  
25%    1.728441e+05  1.734692e+05  
50%    2.317158e+05  2.328665e+05  
75%    3.321367e+05  3.331606e+05  
max    1.607357e+06  1.611650e+06  

[8 rows x 123 columns]

There are 121 float64 (numbers), 2 int64 (intergers), and 2 object (class) column entries. Assuming that the int64 are the Region ID and Size Rank columns, I would need to change its data type. It seems that there aren't any apparent missing values

# Missing Values

In [ ]:
#Identify missing values
df.isnull().sum()

RegionID       0
SizeRank       0
RegionName     0
StateName      0
1/31/2015     40
              ..
9/30/2024      0
10/31/2024     0
11/30/2024     0
12/31/2024     0
1/31/2025      0
Length: 125, dtype: int64

In [ ]:
#View missing values
df[df.isnull().any(axis=1)]

RegionID  SizeRank           RegionName StateName     1/31/2015  \
34     395059        36         San Jose, CA        CA  774847.62690   
91     394549        95           Durham, NC        NC  190235.89840   
118    395244       122             York, PA        PA  155555.43120   
135    394327       139        Anchorage, AK        AK  280916.94920   
205    394763       213     Lake Charles, LA        LA           NaN   
221    394680       229             Hilo, HI        HI           NaN   
251    394428       260       Burlington, NC        NC           NaN   
319    395153       329     The Villages, FL        FL  249417.34530   
337    394810       348        Lumberton, NC        NC           NaN   
387    395190       400         Victoria, TX        TX           NaN   
428    394744       442         Key West, FL        FL           NaN   
443    395085       457           Seneca, SC        SC  136232.55490   
477    394305       493       Alamogordo, NM        NM           NaN   
488    394827       504           Marion, OH        OH   75784.33589   
509    394309       526        Albemarle, NC        NC           NaN   
513    394443       530         Carlsbad, NM        NM           NaN   
551    753920       569           Sonora, CA        CA           NaN   
559    394402       577            Boone, NC        NC  234392.58070   
586    394878       605      Morgan City, LA        LA           NaN   
593    394437       612      Ca-¦on City, CO        CO           NaN   
596    394478       615           Clovis, NM        NM           NaN   
598    394431       617         Cadillac, MI        MI   81942.15274   
613    753916       632        Sandpoint, ID        ID  234471.03640   
629    394650       648       Greenville, MS        MS           NaN   
639    395037       659            Rolla, MO        MO           NaN   
641    753904       661           Marion, NC        NC           NaN   
680    395219       700      West Plains, MO        MO   97061.07097   
690    395203       710           Warren, PA        PA           NaN   
697    395099       717         Sikeston, MO        MO           NaN   
708    394451       728        Centralia, IL        IL           NaN   
722    395137       745  Sulphur Springs, TX        TX           NaN   
746    394536       772       Dodge City, KS        KS           NaN   
747    395211       773        Watertown, SD        SD           NaN   
761    395165       788             Troy, AL        AL           NaN   
777    394473       806        Cleveland, MS        MS           NaN   
795    394673       826             Hays, KS        KS           NaN   
804    753879       836           Elkins, WV        WV           NaN   
810    394642       845           Grants, NM        NM           NaN   
816    394706       851        Indianola, MS        MS           NaN   
821    394643       857       Great Bend, KS        KS           NaN   
847    394793       885          Liberal, KS        KS           NaN   
851    394470       889       Clarksdale, MS        MS           NaN   
854    394718       892        Jamestown, ND        ND           NaN   
856    394656       894          Grenada, MS        MS           NaN   
860    394659       898           Guymon, OK        OK           NaN   
863    394700       901            Huron, SD        SD           NaN   
871    394965       911          Parsons, KS        KS           NaN   
872    395003       912            Price, UT        UT           NaN   
879    753927       924       Winnemucca, NV        NV           NaN   
886    395186       932       Vermillion, SD        SD           NaN   
888    394968       934            Pecos, TX        TX           NaN   

        2/28/2015     3/31/2015     4/30/2015     5/31/2015     6/30/2015  \
34   784297.14230  789912.83140  792572.42870  795334.41140  800136.45380   
91   191264.63500  192205.65430  193044.16210  193711.02550  194248.92900   
118  155596.23040  155684.951

There's missing data on some regions for the 10 yr period. Given that the project focuses on urban areas, I would eliminate most of the regions with "NaN" values on the 1/31/2015 column; with exceptions being Hilo HI, Winnemucca NV  and Morgan City LA due to them having data that rounds up to 10 years.

In [ ]:
#Eliminate rows 205, 251, 337, 387, 428, 477, 509, 513, 551, 593, 596, 629, 639, 641, 690, 697, 708, 722, 746, 747, 761, 777, 795, 804, 810, 816, 821, 847, 851, 854, 856, 860, 863, 871, 872, 886, 888
# Create a list of rows to drop
rows_to_drop = [205, 251, 337, 387, 428, 477, 509, 513, 551, 593, 596, 629, 639, 641, 690, 697, 708, 722, 746, 747, 761, 777, 795, 804, 810, 816, 821, 847, 851, 854, 856, 860, 863, 871, 872, 886, 888]

# Create a boolean mask to identify rows to keep
rows_to_keep = ~df.index.isin(rows_to_drop)

# Filter the DataFrame to keep only the selected rows
df = df[rows_to_keep]


In [ ]:
#View missing values
df[df.isnull().any(axis=1)]

RegionID  SizeRank        RegionName StateName     1/31/2015  \
34     395059        36      San Jose, CA        CA  774847.62690   
91     394549        95        Durham, NC        NC  190235.89840   
118    395244       122          York, PA        PA  155555.43120   
135    394327       139     Anchorage, AK        AK  280916.94920   
216    394680       229          Hilo, HI        HI           NaN   
309    395153       329  The Villages, FL        FL  249417.34530   
461    394827       504        Marion, OH        OH   75784.33589   
518    394402       577         Boone, NC        NC  234392.58070   
542    394878       605   Morgan City, LA        LA           NaN   
551    394431       617      Cadillac, MI        MI   81942.15274   
564    753916       632     Sandpoint, ID        ID  234471.03640   
615    395219       700   West Plains, MO        MO   97061.07097   
748    753927       924    Winnemucca, NV        NV           NaN   

        2/28/2015     3/31/2015     4/30/2015     5/31/2015     6/30/2015  \
34   784297.14230  789912.83140  792572.42870  795334.41140  800136.45380   
91   191264.63500  192205.65430  193044.16210  193711.02550  194248.92900   
118  155596.23040  155684.95120  155999.77230  156345.22320  156654.06880   
135  281264.07590  281201.16310  281848.67690  282884.57480  284045.33820   
216           NaN  301111.26900  301349.50800  301863.17840  302767.57870   
309  250850.80290  251835.86340  252317.73240  252878.79020  253445.53450   
461   75612.56171   75573.22936   75814.57776   76477.13960   77447.84494   
518  234374.58110  233927.45350  233590.95150  234026.42100  234702.08530   
542           NaN           NaN           NaN           NaN  122681.62510   
551   82186.26773   82615.13473   82813.34516   82986.30082   82875.91586   
564  235556.24480  237066.90750  238951.46200  241391.73470  243890.75490   
615   97049.87061   97336.32622   97869.77480   98631.34429   99372.24724   
748           NaN           NaN  169298.38370  170058.78410  170541.27880   

     ...     4/30/2024     5/31/2024     6/30/2024     7/31/2024  \
34   ...  1.500225e+06  1.517969e+06  1.532970e+06  1.547438e+06   
91   ...  4.007210e+05  4.022736e+05  4.023317e+05  4.019929e+05   
118  ...  2.726250e+05  2.736776e+05  2.742858e+05  2.745568e+05   
135  ...  3.760992e+05  3.767975e+05  3.775567e+05  3.781377e+05   
216  ...  5.454551e+05  5.499221e+05  5.525003e+05  5.533802e+05   
309  ...  3.997921e+05  4.004152e+05  4.003704e+05  3.997973e+05   
461  ...  1.572147e+05  1.592281e+05  1.604160e+05  1.609982e+05   
518  ...  4.822397e+05  4.869041e+05  4.897079e+05  4.911535e+05   
542  ...  1.248786e+05  1.240587e+05  1.226406e+05  1.209458e+05   
551  ...  2.048485e+05  2.061474e+05  2.068821e+05  2.075027e+05   
564  ...  6.000503e+05  6.008588e+05  6.011759e+05  6.013480e+05   
615  ...  1.961902e+05  1.968031e+05  1.974118e+05  1.984302e+05   
748  ...  3.005566e+05  3.022979e+05  3.048386e+05  3.060542e+05   

        8/31/2024     9/30/2024    10/31/2024    11/30/2024    12/31/2024  \
34   1.560217e+06  1.572113e+06  1.582338e+06  1.594660e+06  1.607357e+06   
91   4.014338e+05  4.012739e+05  4.012258e+05  4.014216e+05  4.020281e+05   
118  2.750062e+05  2.757556e+05  2.769839e+05  2.785155e+05  2.804334e+05   
135  3.785499e+05  3.791048e+05  3.804684e+05  3.821882e+05  3.842849e+05   
216  5.545653e+05  5.558858e+05  5.566816e+05  5.552871e+05  5.535380e+05   
309  3.991895e+05  3.986546e+05  3.983836e+05  3.982567e+05  3.979536e+05   
461  1.610829e+05  1.611385e+05  1.614373e+05  1.618928e+05  1.627494e+05   
518  4.920046e+05  4.935603e+05  4.948575e+05  4.942818e+05  4.931413e+05   
542  1.194982e+05  1.186907e+05  1.189441e+05  1.202798e+05  1.223594e+05   
551  2.089268e+05  2.096362e+05  2.101955e+05  2.108929e+05  2.118441e+05   
564  6.036330e+05  6.068285e+05  6.103017e+05  6.122630e+05  6.133961e+05   
615  1.995612e+05  1.999053e+05  2.001302e+05  2.013645e+05  2.029485e+05   
748  

The remaining missing values can now be filled through median

In [ ]:
#Replace missing values with median
# Select numeric columns
numeric_cols = df.select_dtypes(include=['number']).columns

# Replace missing values with median for numeric columns
df[numeric_cols] = df[numeric_cols].fillna(df[numeric_cols].median())

In [ ]:
#View missing values
df[df.isnull().any(axis=1)]

Empty DataFrame
Columns: [RegionID, SizeRank, RegionName, StateName, 1/31/2015, 2/28/2015, 3/31/2015, 4/30/2015, 5/31/2015, 6/30/2015, 7/31/2015, 8/31/2015, 9/30/2015, 10/31/2015, 11/30/2015, 12/31/2015, 1/31/2016, 2/29/2016, 3/31/2016, 4/30/2016, 5/31/2016, 6/30/2016, 7/31/2016, 8/31/2016, 9/30/2016, 10/31/2016, 11/30/2016, 12/31/2016, 1/31/2017, 2/28/2017, 3/31/2017, 4/30/2017, 5/31/2017, 6/30/2017, 7/31/2017, 8/31/2017, 9/30/2017, 10/31/2017, 11/30/2017, 12/31/2017, 1/31/2018, 2/28/2018, 3/31/2018, 4/30/2018, 5/31/2018, 6/30/2018, 7/31/2018, 8/31/2018, 9/30/2018, 10/31/2018, 11/30/2018, 12/31/2018, 1/31/2019, 2/28/2019, 3/31/2019, 4/30/2019, 5/31/2019, 6/30/2019, 7/31/2019, 8/31/2019, 9/30/2019, 10/31/2019, 11/30/2019, 12/31/2019, 1/31/2020, 2/29/2020, 3/31/2020, 4/30/2020, 5/31/2020, 6/30/2020, 7/31/2020, 8/31/2020, 9/30/2020, 10/31/2020, 11/30/2020, 12/31/2020, 1/31/2021, 2/28/2021, 3/31/2021, 4/30/2021, 5/31/2021, 6/30/2021, 7/31/2021, 8/31/2021, 9/30/2021, 10/31/2021, 11/30/2021, 12/31/2021, 1/31/2022, 2/28/2022, 3/31/2022, 4/30/2022, 5/31/2022, 6/30/2022, 7/31/2022, 8/31/2022, 9/30/2022, 10/31/2022, 11/30/2022, 12/31/2022, ...]
Index: []

[0 rows x 125 columns]

In [ ]:
#Save dataset
df.to_csv('cleaned_HousePricesdataset.csv', index=False)

# Fix Data Types and Duplicates

In [ ]:
#Re-import dataset
import pandas as pd
df = pd.read_csv('cleaned_HousePricesdataset.csv')

In [ ]:
#Change data type int64 to object
df['RegionID'] = df['RegionID'].astype(object)
df['SizeRank'] = df['SizeRank'].astype(object)

In [ ]:
#Verify if the data types where changed
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 736 entries, 0 to 735
Columns: 125 entries, RegionID to 1/31/2025
dtypes: float64(121), object(4)
memory usage: 718.9+ KB


In [ ]:
#Identify duplicate records
df[df.duplicated()]

Empty DataFrame
Columns: [RegionID, SizeRank, RegionName, StateName, 1/31/2015, 2/28/2015, 3/31/2015, 4/30/2015, 5/31/2015, 6/30/2015, 7/31/2015, 8/31/2015, 9/30/2015, 10/31/2015, 11/30/2015, 12/31/2015, 1/31/2016, 2/29/2016, 3/31/2016, 4/30/2016, 5/31/2016, 6/30/2016, 7/31/2016, 8/31/2016, 9/30/2016, 10/31/2016, 11/30/2016, 12/31/2016, 1/31/2017, 2/28/2017, 3/31/2017, 4/30/2017, 5/31/2017, 6/30/2017, 7/31/2017, 8/31/2017, 9/30/2017, 10/31/2017, 11/30/2017, 12/31/2017, 1/31/2018, 2/28/2018, 3/31/2018, 4/30/2018, 5/31/2018, 6/30/2018, 7/31/2018, 8/31/2018, 9/30/2018, 10/31/2018, 11/30/2018, 12/31/2018, 1/31/2019, 2/28/2019, 3/31/2019, 4/30/2019, 5/31/2019, 6/30/2019, 7/31/2019, 8/31/2019, 9/30/2019, 10/31/2019, 11/30/2019, 12/31/2019, 1/31/2020, 2/29/2020, 3/31/2020, 4/30/2020, 5/31/2020, 6/30/2020, 7/31/2020, 8/31/2020, 9/30/2020, 10/31/2020, 11/30/2020, 12/31/2020, 1/31/2021, 2/28/2021, 3/31/2021, 4/30/2021, 5/31/2021, 6/30/2021, 7/31/2021, 8/31/2021, 9/30/2021, 10/31/2021, 11/30/2021, 12/31/2021, 1/31/2022, 2/28/2022, 3/31/2022, 4/30/2022, 5/31/2022, 6/30/2022, 7/31/2022, 8/31/2022, 9/30/2022, 10/31/2022, 11/30/2022, 12/31/2022, ...]
Index: []

[0 rows x 125 columns]

There's no duplicate records on the dataset. Additionally, I've changed the int64 data types to object

In [ ]:
#Save file
df.to_csv('cleaned_HousePrices_dataset.csv', index=False)

# Outliers

In [ ]:
#Import new file name
import pandas as pd
df = pd.read_csv('cleaned_HousePrices_dataset.csv')

In [ ]:
#Identify outliers in housing prices
# Select the date columns (assuming they start with '1/31/2015')
date_columns = df.columns[df.columns.get_loc('1/31/2015'):]

# Melt the DataFrame to create a single 'Price' column
melted_df = pd.melt(df,
                   id_vars=['RegionName', 'StateName'],  # Keep 'RegionName' and 'StateName' as identifiers
                   value_vars=date_columns,  # Use the date columns for values
                   var_name='Date',
                   value_name='Price')

# Create box plots for each state
fig = px.box(melted_df, x='StateName', y='Price', points="all")
fig.update_layout(title_text='Housing Prices by State')
fig.show()

In [ ]:
#Make the boxplots clearer
# Select the date columns (assuming they start with '1/31/2015')
date_columns = df.columns[df.columns.get_loc('1/31/2015'):]

# Melt the DataFrame to create a single 'Price' column
melted_df = pd.melt(df,
                   id_vars=['RegionName', 'StateName'],  # Keep 'RegionName' and 'StateName' as identifiers
                   value_vars=date_columns,  # Use the date columns for values
                   var_name='Date',
                   value_name='Price')

# Create box plots for each state
fig = px.box(melted_df, x='StateName', y='Price', points="all")

# Update layout for better visualization
fig.update_layout(
    width=1500,  # Increase width of the plot
    height=600,  # Increase height, if needed
    title_text='Housing Prices by State'
)

# Increase marker size for outliers (optional)
fig.update_traces(marker=dict(size=4))  # Adjust size as needed

fig.show()

The box plots were created to show the pricing per state with no regards to the dates. This allows us to better see any vast differences in pricing between the 10 years (represented by those with large outliers).

The dots on the side of the boxplots represent all of the data thats within each state. We can identify outliers by seeing those that aren't aligned with the box plots themselves. In this case, we can see that outliers are quite prominent in states such as WY, NY, CA, and CO. However, I will be keeping the outliers.

# National House Prices Trend

It is helpful to know the general house pricing trend. Therefore, this section will be attempting to answer: What's the national housing price per year?

In [ ]:
#Create a graph that represents national average housing price per year
# Assuming your DataFrame is named 'df' and the date columns start with '1/31/2015'
date_columns = df.columns[df.columns.get_loc('1/31/2015'):]

# Calculate the average price for each date (year)
average_prices = df[date_columns].mean()

# Create a new DataFrame for the average prices
average_prices_df = pd.DataFrame({'Date': average_prices.index, 'AveragePrice': average_prices.values})

# Create the line graph
fig = px.line(average_prices_df, x='Date', y='AveragePrice', title='National Average Housing Price per Year')

# Customize the graph (optional)
fig.update_layout(xaxis_title='Year', yaxis_title='Average Housing Price')

# Show the graph
fig.show()


The line chart represents the national house prices average per year. It gives us an idea on how the nation is performing house-pricing wise, and allows us to see a trend on how housing price is increasing. We can see that there was a sharp increase in house prices around Oct 2020, and it kept (generally speaking) increasing since then. This data also helps in identifying whether a state/region is below/above national data; and for any further analysis which can benefit from a general data reference